### PR TITLE
feat: label expression parser

### PR DESCRIPTION
**Pull request by nklaassen**
This PR builds on https://github.com/gravitational/teleport/pull/26177 to implement the label expression parser as defined in [RFD 116](https://github.com/gravitational/teleport/blob/master/rfd/0116-label-expressions.md#expression-syntax).

### CODE REVIEW

File: lib/services/label_expressions.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+import (
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/gravitational/trace"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
```
Comment (codingllama) [inline lib/services/label_expressions.go:25]: Double import for logrus.

File: lib/services/label_expressions.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+func parseLabelExpression(expr string) (labelExpression, error) {
+	if parsedExpr, ok := labelExpressionCache.Get(expr); ok {
+		return parsedExpr, nil
+	}
+	parsedExpr, err := labelExpressionParser.Parse(expr)
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing label expression")
+	}
+	if evicted := labelExpressionCache.Add(expr, parsedExpr); evicted {
+		logrus.Warn("evicting entry from label expression cache")
```
Comment (codingllama) [inline lib/services/label_expressions.go:48]: nit: Capitalize log messages.

```suggestion
		logrus.Warn("Evicting entry from label expression cache")
```

Also, is this worth a warning?

File: lib/services/label_expressions.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+const (
+	cacheSizeEnvVar  = "TELEPORT_EXPRESSION_CACHE_SIZE"
+	defaultCacheSize = 1000
+)
+
+func mustNewLabelExpressionCache() *lru.Cache[string, labelExpression] {
+	cacheSize := defaultCacheSize
+	if env := os.Getenv(cacheSizeEnvVar); env != "" {
+		if envCacheSize, err := strconv.ParseUint(env, 10, 0); err != nil {
+			log.WithError(err).Warn("parsing " + cacheSizeEnvVar)
```
Comment (codingllama) [inline lib/services/label_expressions.go:68]:
```suggestion
			log.WithError(err).Warnf("Parsing %v", cacheSizeEnvVar)
```

File: lib/services/label_expressions.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+			"user.spec.traits": typical.DynamicVariable(
+				func(env labelExpressionEnv) (map[string][]string, error) {
+					return env.userTraits, nil
+				}),
+			"resource.metadata.labels": typical.DynamicMapFunction(
+				func(env labelExpressionEnv, key string) (string, error) {
+					label, _ := env.resourceLabelGetter.GetLabel(key)
+					return label, nil
+				}),
+			"labels": typical.DynamicMapFunction(
```
Comment (codingllama) [inline lib/services/label_expressions.go:92]: Is "labels" an alias for "resource.metadata.labels"? Maybe use the same DynamicMapFunction instance to make it clearer?

File: lib/services/label_expressions.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+	labelExpressionCache  = mustNewLabelExpressionCache()
+	labelExpressionParser = mustNewLabelExpressionParser()
+)
+
+const (
+	cacheSizeEnvVar  = "TELEPORT_EXPRESSION_CACHE_SIZE"
+	defaultCacheSize = 1000
+)
+
+func mustNewLabelExpressionCache() *lru.Cache[string, labelExpression] {
```
Comment (codingllama) [inline lib/services/label_expressions.go:64]: Suggestion: make "must" function call the "regular" function, so we have both the must and the error-returning variant? For example, "mustNewLabelExpressionCache" calls "newLabelExpressionCache", and the latter returns an error.

File: lib/services/label_expressions.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+	}
+	if evicted := labelExpressionCache.Add(expr, parsedExpr); evicted {
+		logrus.Warn("evicting entry from label expression cache")
+	}
+	return parsedExpr, nil
+
+}
+
+var (
+	labelExpressionCache  = mustNewLabelExpressionCache()
```
Comment (codingllama) [inline lib/services/label_expressions.go:55]: What is the reasoning behind having a label expressions cache?

Is the parsing expensive?

Which sources of expressions do we expect the cache to cover?

File: lib/utils/replace.go [hunk 7]
```diff
@@ -128,6 +128,22 @@ func SliceMatchesRegex(input string, expressions []string) (bool, error) {
 	return false, nil
 }
 
+func RegexMatchesSlice(inputs []string, expression string) (bool, error) {
```
Comment (codingllama) [inline lib/utils/replace.go:131]: Add godoc and a few tests?

Maybe call it `RegexMatchesSliceAny` or `RegexMatchesAny`?

File: lib/services/label_expressions.go (hunk 6 above)
Comment (nklaassen) [reply lib/services/label_expressions.go:55]: https://github.com/gravitational/teleport/blob/master/rfd/0116-label-expressions.md#performance

Parsing is usually quite expensive compared to evaluation. My main performance concern is operations like `ListNodes` where we check access to each and every node in the cluster. If you have a role with a simple expression like `labels["env"] == "staging"`, and you call `ListNodes` with 50k nodes, without a cache we would be parsing that expression 50k times. With a cache we only parse it once, after that it's a simple string comparison.

I ran a couple benchmarks without the cache and ListNodes got like 2-3x slower. With the cache, `node_labels_expression: 'labels["env"] == staging'` looks like it's almost 30% faster than
```
node_labels:
  env: staging
```

This cache will cover all "label expressions", which will show up in role specs, `node_labels_expression`, `app_labels_expression`, etc...

Comment (codingllama) [reply lib/services/label_expressions.go:55]: Gotcha, thanks for the explanation.

File: lib/services/label_expressions.go (hunk 4 above)
Comment (nklaassen) [reply lib/services/label_expressions.go:100]: Thanks for raising this. It was meant to be an alias, but I had another look at `resource.metadata.labels` vs `labels` in the [resource filters docs](https://goteleport.com/docs/reference/predicate-language/#resource-filtering) and realized that that former is only the static labels while the latter is combined with the dynamic labels. I think for this we'll always want to use the combined so I've removed the `resource.metadata.labels` variable.

**Review by gabrielcorado — APPROVED** [review]

**Review by codingllama — APPROVED** [review]
Apologies for the delay in the 2nd pass. LGTM.

**Pull request merged**